### PR TITLE
Feat/console virtualization

### PR DESCRIPTION
**Pull request by johann-crabnebula**
# TL;DR

This PR aims to improve the performance of the console tab. The goal was to at least be able to confidently display 10k logs and make the experience as smooth as possible.  The dependency `tanstack-virtual` for `solidjs` was added to make sure we only render the DOM elements which are necessary for the view. 

## Acceptance criteria

- [ ] General handling of 10k + logs
- [ ] Switching between tabs without breaking (for example from console -> sources -> console / note that the calls tab is currently not working with a large number of spans, so please try to test the other tabs when switching, PR for the calls tab is coming)
- [ ] Filtering should be smooth with logs coming in and when inactive
- [ ] Console tab should work fine with cached logs when disconnected

## Let me know

If there is anything odd in general please let me know.

### Help with testing this

In my testing I just spammed events from my tauri app. Please feel free to be more creative with testing:

```rs
let mut count = 0;
// Infinite loop
loop {
    count += 1;
    tokio::time::sleep(Duration::from_millis(1)).await;
    window
        .emit(
            format!("test{}-event", count).as_str(),
            &EventPayload {
                key: "count",
                value: count.to_string(),
            },      
    ).unwrap();
    tracing::debug!("test{}-event", count);

    if count == 15000 {
        // Exit this loop
        break;
    }
}
```

Fixes DT-112

### CODE REVIEW

**Review by CrabNejonas — APPROVED** [review]
Looks good, not gonna argue with you about any JS stuff, but perfand usability is good 👍

**Review by CrabNejonas — DISMISSED** [review]

**Review by denjell-crabnebula — DISMISSED** [review]
Johann and I talked this through, and it seems good enough for now. I think we should ship it and then focus on speed / perf enhancements for the Tauri app version.

File: clients/web/src/components/console/log-event-entry.tsx [hunk 1]
```diff
@@ -0,0 +1,75 @@
+import { Show, untrack } from "solid-js";
+import { getLogMetadata } from "~/lib/console/get-log-metadata";
+import { timestampToDate, formatTimestamp } from "~/lib/formatters";
+import { getLevelClasses } from "~/lib/console/get-level-classes";
+import { useMonitor } from "~/context/monitor-provider";
+import { processFieldValue } from "~/lib/span/process-field-value";
+import type { LogEvent } from "~/lib/proto/logs";
+import clsx from "clsx";
```
Comment (atila-crabnebula) [inline clients/web/src/components/console/log-event-entry.tsx:8]: `clsx` is actually unnecessary.
Solid's JSX has a `classList` attribute that works the exact same way.

File: clients/web/src/components/console/log-event-entry.tsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+          {(logTarget) => (
+            <span class="text-slate-400 group-hover:text-slate-100 transition-colors">
+              {logTarget()}
+            </span>
+          )}
+        </Show>
+        <Show when={metadata?.location?.file}>
+          {(filePath) => (
+            <>
+              {getFileNameFromPath(filePath())}:{metadata?.location?.line ?? ""}
```
Comment (atila-crabnebula) [inline clients/web/src/components/console/log-event-entry.tsx:68]: this would be better to handle the non-existing line within `getFileNameFromPath` and omitting the DOM node completely instead of printing an empty string. So the method returns the line or a falsy value (`null`, perhaps).

So the signature looks like:

```jsx
<Show when={getFileNameFromPath(metadata?.location?.file)>
  {line => (<span>{line()</span>)}
</Show>
```
wdyt?

File: clients/web/src/lib/connection/transport.tsx [hunk 3]
```diff
@@ -121,12 +121,9 @@ export function addStreamListneners(
...
     const spansUpdate = update.spansUpdate;
     if (spansUpdate && spansUpdate.spanEvents.length > 0) {
-      setMonitorData(
-        "spans",
-        produce((clonedSpans) =>
-          updatedSpans(clonedSpans, spansUpdate.spanEvents)
-        )
-      );
+      setMonitorData("spans", (spans) => [
+        ...updatedSpans(spans, spansUpdate.spanEvents),
```
Comment (atila-crabnebula) [inline clients/web/src/lib/connection/transport.tsx:125]: this is mutating the entire array. Every time this `setMonitorData` is called, we replace the entire array. How about wrapping it in a `reconcile()` so only the needed keys are mutated? 🤔

File: clients/web/src/views/dashboard/console.tsx [hunk 4]
```diff
@@ -23,12 +19,48 @@ export default function Console() {
     levels: [0, 1, 2, 3, 4],
   });
   const [filter, setFilter] = createStore<LogFilterObject>(initialFilters());
-  const filteredLogs = () => filterLogs(monitorData, filter, monitorData.logs);
+
+  const [filteredLogs, setFilteredLogs] = createSignal<LogEvent[]>([]);
```
Comment (atila-crabnebula) [inline clients/web/src/views/dashboard/console.tsx:23]: not sure I follow all the `untrack` happening in this code. Could you explain?
I'm thinking maybe this `createSignal` should be a store so we don't need to bail out so much...

**Review by atila-crabnebula — CHANGES_REQUESTED** [review]
I have a few questions, so I'll wait for Johann to be back before merging this.

File: clients/web/src/components/console/log-event-entry.tsx (hunk 2 above)
Comment (johann-crabnebula) [reply clients/web/src/components/console/log-event-entry.tsx:68]: I will fix this.

File: clients/web/src/lib/connection/transport.tsx (hunk 3 above)
Comment (johann-crabnebula) [reply clients/web/src/lib/connection/transport.tsx:125]: I will try to use Solid's `reconcile` here

File: clients/web/src/views/dashboard/console.tsx (hunk 4 above)
Comment (johann-crabnebula) [reply clients/web/src/views/dashboard/console.tsx:23]: As explained in person this is meant to only trigger the effects based on changes to the corresponding reactive property.

File: clients/web/src/components/console/log-event-entry.tsx (hunk 2 above)
Comment (johann-crabnebula) [reply clients/web/src/components/console/log-event-entry.tsx:68]: Wrote a `getFileLineFunction` that either returns a fully fledged line or null:

```typescript
import { Location } from "../proto/common";
import { getFileNameFromPath } from "./get-file-name-from-path";

export function getFileLineFromLocation(location: Location | undefined) {
  if (!location || !location.file) return null;

  let line = getFileNameFromPath(location.file);

  if (location.line) line += `:${location.line}`;

  return line;
}
```

Template now looks like this:

```jsx
<Show when={getFileLineFromLocation(metadata?.location)}>
  {(line) => <span>{line()}</span>}
</Show>
```

File: clients/web/src/lib/connection/transport.tsx (hunk 3 above)
Comment (johann-crabnebula) [reply clients/web/src/lib/connection/transport.tsx:125]: For the use of reconcile the `updatedSpans` function and the receiving piece on the calls tab might need updates. I would like to postpone this change to the calls virtualization PR.

**Review by atila-crabnebula — CHANGES_REQUESTED** [review]
something is going on at the spans tab.

## Bug

1. Run tauri-v1 example
2. Open the spans tab
3. Fire a span test

It will never catch the end event.

**Comment by johann-crabnebula** [comment]
Reverted the change for the spans. It will slow down the overall performance but this unblocks the PR, so we can move on.

**Review by atila-crabnebula — APPROVED** [review]
